### PR TITLE
Add option to set build timeout with build argument

### PR DESCRIPTION
**Pull request by mlahargou**
Coverage takes more than our default timeout of two hours which is set in 
the cimpler config. Add the ability to set the build timeout using a build
command argument.

Related: https://github.com/iFixit/ifixit/pull/35075

@danielbeardsley @BaseInfinity

### CODE REVIEW

**Comment by BaseInfinity** [comment]
CR 👙

File: plugins/git-build.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ function buildConsumer(config, cimpler, repoPath) {
             BUILD_STATUS: build.status,
             BUILD_QUEUED_AT: build.queuedAt,
          },
-         timeout: config.timeout || 0,
+         timeout: process.env.buildTimeout || config.timeout || 0,
```
Comment (danielbeardsley) [inline plugins/git-build.js:38]: Hmmm...  I think this won't help cause `process.env` here is the cimpler process. We alread have a way to specify a timeout for the plugin, what we want is a way to specify the timeout for a particular build.

Perhaps provide a `--timeout` argument to the `build` command and have it sent as `build.timeout`

Comment (mlahargou) [reply plugins/git-build.js]: Is that what you were thinking? https://github.com/danielbeardsley/cimpler/pull/104/commits/402110fdd07d4642cb8aef2af6259e1f06f5960c

**Comment by BaseInfinity** [comment]
CR 🚴‍♀️  looks good to me

File: bin/cimpler [hunk 2]
```diff
@@ -39,6 +39,10 @@ args           = require('optimist')
    .options('port', {
       alias: 'p',
       describe: 'HTTP port of the cimpler server (defaults to the value in config)'
+   })
+   .options('timeout', {
+      alias: 't',
+      describe: 'Specify the timeout of the build (in milliseconds)'
```
Comment (addison-grant) [inline bin/cimpler:45]: Somewhat minor, but I think this wording could be improved
```suggestion
      describe: 'Specify the timeout for builds (in milliseconds)'
```

**Comment by danielbeardsley** [comment]
CR 👍   Nice!

**Comment by mlahargou** [comment]
Looks like a flaky test!
https://travis-ci.org/github/danielbeardsley/cimpler/builds/742834382

**Comment by addison-grant** [comment]
> Looks like a flaky test!
> https://travis-ci.org/github/danielbeardsley/cimpler/builds/742834382

I think that was just an aborted build due to a new commit? 

> `This job was cancelled because the "Auto Cancellation" feature is currently enabled, and a more recent build (#400) for pull request #104 came in while this job was waiting to be processed.`

image

**Comment by mlahargou** [comment]
@addison-grant, how do we want to go about QA on this?  I guess we can talk about it in QA standup.

**Comment by addison-grant** [comment]
I would say either:
1. spin up a cimpler and test running 
2.

**Comment by mlahargou** [comment]
> write some sort of end-to-end automated test

I wrote a little test that makes sure the timeout argument gets attached to the build.  This isn't exactly end to end, but I think it is a little helpful.

File: test/cli.test.js [hunk 3]
```diff
@@ -109,6 +109,33 @@ describe("CLI server command", function() {
...
+         httpPort: httpPort,
+         testMode: true  // Don't console.log() anything
+      });
+      var timeout = 1 * 1000; // 1 second timeout
+      cimpler.on('buildFinished', function(build) {
+         assert.equal(build.buildTimeout, timeout);
+      });
+      var args = ["-h", "127.0.0.1", "-p", httpPort];
+      var proc = exec(bin,  args.concat(["build", '--timeout=' + timeout]));
+      cimpler.consumeBuild(function(inBuild, started, finished) {
```
Comment (addison-grant) [inline test/cli.test.js:130]: Do we use this convention at all? `_` to indicate the argument isn't used?

```suggestion
      cimpler.consumeBuild(function(_, started, finished) {
```

File: test/cli.test.js [hunk 4]
```diff
@@ -109,6 +109,33 @@ describe("CLI server command", function() {
...
+      var cimpler = new Cimpler({
+         plugins: {
+            cli: {
+            }
+         },
+         httpPort: httpPort,
+         testMode: true  // Don't console.log() anything
+      });
+      var timeout = 1 * 1000; // 1 second timeout
+      cimpler.on('buildFinished', function(build) {
```
Comment (addison-grant) [inline test/cli.test.js:125]: Just for my understanding, this could have been `'buildAdded'` or `'buildStarted'`?

**Comment by addison-grant** [comment]
> I wrote a little test that makes sure the timeout argument gets attached to the build. This isn't exactly end to end, but I think it is a little helpful.

Looks good to me! Did you confirm that an error in the git-build plugin timeout logic causes the test to fail?

File: test/cli.test.js (hunk 4 above)
Comment (danielbeardsley) [reply test/cli.test.js:125]: I think you're right. This proves the argument makes it through... but not that timeout is respected,

File: test/cli.test.js (hunk 3 above)
Comment (mlahargou) [reply test/cli.test.js:130]: I've never used it 🤷

File: test/cli.test.js (hunk 4 above)
Comment (mlahargou) [reply test/cli.test.js:125]: Yeah, I was having a hard time testing that the timeout is respected. I would be happy to collab and maybe get a test written for that if either of you wanted.

Comment (mlahargou) [reply test/cli.test.js:125]: Idk how important that test would be.

**Comment by mlahargou** [comment]
Yup!

```
1) CLI build timeout
     sets the build timeout:

    Uncaught AssertionError [ERR_ASSERTION]: 1000 == 1001
    + expected - actual

    -1000
    +1001

    at Cimpler.<anonymous> (test/cli.test.js:127:17)
    at /Users/michaellahargou/ifixit/cimpler/lib/cimpler.js:90:24
    at processTicksAndRejections (internal/process/task_queues.js:79:11

File: test/cli.test.js (hunk 4 above)
Comment (danielbeardsley) [reply test/cli.test.js:125]: I think it's important to test that both the setting is set and that the setting is used. I added a test!

**Comment by danielbeardsley** [comment]
CR 👍

**Comment by mlahargou** [comment]
Thanks, @danielbeardsley!  CR :on-the-test:

**Comment by danielbeardsley** [comment]
OMG, Thank you!!!   

CR 👍

**Comment by addison-grant** [comment]
`QAing`

**Comment by danielbeardsley** [comment]
I don't think QA is worth it here. We have robust tests and the cost if we got it wrong is super low.

**Pull request merged**